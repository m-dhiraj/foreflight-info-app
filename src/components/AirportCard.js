import React from 'react';
import { useDispatch } from 'react-redux';
import { removeAirport } from '../redux/airportSlice';
import { celsiusToFahrenheit, knotsToMPH, getWindDirection, getCloudCoverageDescription, calculateTimeOffset } from '../functions/helperFunctions';
import '../styles/AirportCard.css';

const AirportCard = ({ airportValue, airportCode }) => {
    // Handle removing airport by using removeAirport reducer
    const dispatch = useDispatch();
    const handleRemove = () => {
        dispatch(removeAirport(airportCode));
    };
    // Only access the relevant data from the passed prop
    const airport = airportValue.airportData;
    const weather = airportValue.weatherData.report.conditions;
    const forecastConditions = airportValue.weatherData.report.forecast.conditions;
    const relevantConditions = forecastConditions.slice(1, 3);
    // Use imported helper functions to convert raw values to usable outputs
    const tempInFahrenheit = celsiusToFahrenheit(weather.tempC);
    const windSpeedMPH = knotsToMPH(weather.wind.speedKts);
    const windDirection = getWindDirection(weather.wind.direction);
    const cloudCoverage = getCloudCoverageDescription(weather.cloudLayers[0].coverage); 

    // Correctly load the component with the appropriate airport values
    return (
        <div className="airport-card">
            <h1>{airport.icao}</h1>
            <h2>{airport.name}</h2>
            <p>{airport.city}, {airport.state}</p>
            <p>Available Runways:</p>
            <ul>
            {airport.runways.map(runway => (
                <li key={runway.ident}>{runway.ident} - Length: {runway.length} feet</li>
            ))}
            </ul>
            <p>Latitude: {airport.latitude}</p>
            <p>Longitude: {airport.longitude}</p>
            <h2>Current Weather</h2>
            <p>Temperature: {tempInFahrenheit.toFixed(1)} F</p>
            <p>Relative Humidity: {weather.relativeHumidity}%</p>
            <p>Cloud Coverage: {cloudCoverage}</p>
            <p>Visibility: {weather.visibility.prevailingVisSm} miles</p>
            <p>Wind: {windSpeedMPH.toFixed(1)} MPH at {windDirection} degrees</p>
            <h2>Forecast</h2>
            {relevantConditions.map((entry, index) => (
            <div key={index} className="forecast-card">
                <h3>{calculateTimeOffset('2024-02-29T08:00:00', entry.period.dateStart)}</h3>
                <p>Wind Speed: {entry.wind ? knotsToMPH(entry.wind.speedKts).toFixed(1) : 'N/A'} MPH</p>
                <p>Wind Direction: {entry.wind.direction}&deg;</p>
            </div>
            ))}
            <button className="remove-button" onClick={handleRemove}>Remove</button>
        </div>
    );
};

export default AirportCard;
