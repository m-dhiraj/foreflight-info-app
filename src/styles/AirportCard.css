.airport-card {
    background-color: #f9f9f9;
    border-radius: 8px;
    border: 1px solid #ddd;
    padding: 20px;
    margin: 20px auto;
    width: 80%;
    max-width: 1000px;
    font-family: Arial, sans-serif;
}

.airport-card h1, .airport-card h2 {
    color: #333;
}

.airport-card p, .airport-card li {
    color: #666;
    font-size: 16px;
    line-height: 1.5;
}

.airport-card ul {
    list-style-type: none;
    padding: 0;
}

.remove-button {
    background-color: red;
    color: white;
    border: none;
    border-radius: 4px;
    padding: 10px 20px;
    cursor: pointer;
    font-size: 16px;
    transition: background-color 0.3s;
}

.remove-button:hover {
    background-color: darkred;
}

.airport-card h2, .airport-card p {
    margin-top: 5px;
    margin-bottom: 10px;
}

.airport-card > button {
    margin-top: 20px;
    display: block;
    width: 100%;
}
