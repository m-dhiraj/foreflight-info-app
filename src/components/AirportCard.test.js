import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import AirportCard from './AirportCard';
import airportReducer from '../redux/airportSlice';
import airportData from '../data/KAUS-Info.json'; // Default data for KAUS
import weatherData from '../data/KAUS-Weather.json'; // Default weather for KAUS

const store = configureStore({
  reducer: {
    airports: airportReducer
  }
});

// Use Static values from the json files for test
const airportValue = {
  airportData: airportData,
  weatherData: weatherData
};
// Simple snapshot test to test the UI
describe('AirportCard Component', () => {
  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AirportCard airportValue={airportValue} airportCode="KAUS" />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
