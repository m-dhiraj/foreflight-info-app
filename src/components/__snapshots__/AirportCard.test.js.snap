// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirportCard Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="airport-card"
  >
    <h1>
      KAUS
    </h1>
    <h2>
      Austin-Bergstrom International
    </h2>
    <p>
      Austin, Texas
    </p>
    <p>
      Available Runways:
    </p>
    <ul>
      <li>
        18L-36R - Length: 9000 feet
      </li>
      <li>
        18R-36L - Length: 12250 feet
      </li>
    </ul>
    <p>
      Latitude: 30.19452702415562
    </p>
    <p>
      Longitude: -97.66987607813589
    </p>
    <h2>
      Current Weather
    </h2>
    <p>
      Temperature: 48.2 F
    </p>
    <p>
      Relative Humidity: 58%
    </p>
    <p>
      Cloud Coverage: Clouds are Overcast
    </p>
    <p>
      Visibility: 10 miles
    </p>
    <p>
      Wind: 8.1 MPH at NE degrees
    </p>
    <h2>
      Forecast
    </h2>
    <div
      class="forecast-card"
    >
      <h3>
        06:00
      </h3>
      <p>
        Wind Speed: 8.1 MPH
      </p>
      <p>
        Wind Direction: 40°
      </p>
    </div>
    <div
      class="forecast-card"
    >
      <h3>
        12:00
      </h3>
      <p>
        Wind Speed: 4.6 MPH
      </p>
      <p>
        Wind Direction: 20°
      </p>
    </div>
    <button
      class="remove-button"
    >
      Remove
    </button>
  </div>
</DocumentFragment>
`;
